### PR TITLE
Space - Nora

### PR DESCRIPTION
**Pull request by thenora**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a custom method to sort the works by the category
Describe how you approached testing that model method. What edge cases did you come up with? | I tested to make sure each category returned an example fixture of that same type, and for an edge case if a category is empty, it should return an empty array. The view checks if the array is empty, and will display a message encouraging the user to add a new work to the category.
What are `session` and `flash`? What is the difference between them? | session is used to store personalized information related to a user, and flash saves a short-term message related to the success or failure of an interaction. Flash is automatically reset during either the current request (now) or during the next request cycle. Session lasts at least until a user closes the browser, or logs out.
What was one thing that you gained more clarity on through this assignment? | SO MANY THINGS. Fixtures were big. Deploying to heroku was also a little tricky due to some migration errors.
What is the Heroku URL of your deployed application? | https://thenora-media.herokuapp.com/

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+/* Custom bootstrap variables must be set or imported *before* bootstrap. */
+
+// TODO - clean up and add commenting to styling
+
+@import "bootstrap";
+
+@import "**/*";
+
+*,*::before,*::after {
+  box-sizing: border-box
```
Comment (kaidamasaki) [inline app/assets/stylesheets/application.scss:24]: Wow, you _strongly_ prefer `border-box` sizing, huh?

File: app/assets/stylesheets/application.scss [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+  flex-grow: 1;
+  text-align: center
+}
+
+.user-votes__container {
+  margin-top: 2rem
+}
+
+.user-votes__header {
+  color: black
```
Comment (kaidamasaki) [inline app/assets/stylesheets/application.scss:107]: Generally speaking one should avoid color names because they can have browser compatibility issues.  (`black` and `white` don't but it's still generally discouraged.)

File: app/assets/stylesheets/application.scss (hunk 1 above)
Comment (kaidamasaki) [reply app/assets/stylesheets/application.scss:24]: Also, even thought the trailing `;` is optional on the last entry it's strongly encouraged since it makes reordering and adding new rules easier.

File: app/assets/stylesheets/application.scss [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+  background-color: #B2DFDB;
+  margin-bottom: 0.5rem;
+  padding: 2rem 1rem 0.5rem 1rem;
+  text-align: center;
+}
+
+.app-header__header h1 {
+  font-size: 2.5rem;
+	font-weight: 700;
+  margin: 25px auto 40px auto
```
Comment (kaidamasaki) [inline app/assets/stylesheets/application.scss:226]: Indentation.

File: app/controllers/votes_controller.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        user_id: user.id, 
+        work_id: work.id
+      )
+        if upvote.save 
+          flash[:success] = 'Successfully upvoted!'
+        else
+          flash[:error] = 'You already voted for this work'
+        end
+      end
+    redirect_back(fallback_location: root_path)
```
Comment (kaidamasaki) [inline app/controllers/votes_controller.rb:26]: Nice use of `redirect_back`!

File: app/controllers/works_controller.rb [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    @albums = Work.sort_albums[0..9]
+    @spotlight = Work.spotlight
+  end
+
+  def show
+    @work = Work.find_by(id: params[:id])
+    if @work.nil?
+      head :not_found
+      return
+    end
```
Comment (kaidamasaki) [inline app/controllers/works_controller.rb:24]: You do this in a number of places.  In the future you should DRY things like this up by creating a controller filter.

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+</header>
+
+  <section class="alert__container">
+    <% flash.each do |name, message| %>
+      <div class="alert alert-<%= name %>" >
+        <span>
+          <%= message %>
+        </span>
+      <% end %>
```
Comment (kaidamasaki) [inline app/views/layouts/application.html.erb:59]: Indentation.

**Review by kaidamasaki — COMMENTED** [review]
Great job!  Your code was generally quite clean and concise.

Here are a few tips for improving future projects.  😃

**Comment by kaidamasaki** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | No top ten.
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | No filter to find a work.
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention |
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation |
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
